### ISSUE TITLE
Add YearQuarter.of(Year, int) and YearWeek.of(Year, int)

### ISSUE THREAD

**Issue by kluever** [ThreeTen/threeten-extra#155]
Despite `java.time.Year` not getting much love in `java.time` itself (there are 0 public APIs which accept the type), any reason not to add a `Year`-based overload to the static factories in `YearQuarter` and `YearWeek` (analogous to what `LocalDate` does for `Month`: `LocalDate.of(int, Month, int)` and `LocalDate.of(int, int, int)`)?

(In general, we like to require the use of strongly-typed time objects, so `Year` over `int`. Doing so has a lot of benefits, but also ensures you can't screw up the parameter order when calling the static factory methods like `YearQuarter.of(int, int)` or `YearWeek.of(int, int)`).

**Comment by jodastephen** [comment]
A key reason the `Year` class exists is as a starting point for builder-like construction:

    Year.of(2020).atMonth(FEBRUARY).atDay(3);

In that way, `Year` works with `MonthDay` and `Month`.

Since the JDK cannot mention the ThreeTen-Extra types in the same way, it makes sense to add the methods you propose in `YearWeek` and `YearQuarter`.

Alternatively, a different approach could be taken (one I've previously considered). If a class `TemporalFactories` was created (name TBD), it could have factory methods for **all** temporal classes across java.time and ThreeTen-Extra. The crucial difference would be that such a class would be designed for static imports:

    public class TemporalFactories {
      public LocalDate date(int,int,int);
      public LocalDate date(int,Month,int);
      public LocalDate date(int,MonthDay);
      public LocalDate date(int,int);
      public LocalDate date(Year,int,int);
      public LocalDate date(Year,Month,int);
      public LocalDate date(Year,MonthDay);
      public LocalDate date(Year,int);
      public LocalDate date(String);  // parse
      // and so on for other types
    }
    // usage
    callMethod(date(2020, 6, 12));

WDYT?

**Comment by kluever** [comment]
FWIW, we may want to split the discussion of `TemporalFactories` into another issue, but my initial inclination is that I don't see much benefit of:
```
callMethod(date(2020, 6, 12));
```
over
```
callMethod(LocalDate.of(2020, 6, 12));
```

The only real reason to introduce a static factory API is for overloads left out of the JDK. We could also raise the `Year` and `MonthDay`-based overloads to the JDK maintainers - I don't see why those would be contentious to add directly to `LocalDate`.

The ones that obviously couldn't be added, would be the APIs w/ threetenextra types (e.g., with `DayOfMonth`). In that case, I still think I'd rather see folks use the fluent API (`dayOfMonth.atYearMonth(YearMonth)` or `dayOfMonth.atMonth(Month).atYear(year)`).

(Speaking of which, that's _another_ place where a `java.time.Year` overload could be added: `MonthDay.atYear(int)`...I really should write up a proposal for improving `Year` support in `java.time`.)

**Comment by jodastephen** [comment]
OK, well I'm happy to see a PR to add the `Year` factory methods to `YearWeek` and `YearQuarter`.

And separately, I'm happy to see a proposal to OpenJDK on using `Year`.

**Comment by kluever** [comment]
Related to this request, `YearQuarter` has 3 getters:
```java
public int getYear() { ... }
public Quarter getQuarter() { ... }
public int getQuarterValue() { ... }
```

Ideally, we'd have:
```java
public Year getYear() { ... }
public int getYearValue() { ... }
public Quarter getQuarter() { ... }
public int getQuarterValue() { ... }
```

However, that would be a breaking change.  Perhaps we want to add something like this though?
```java
public Year getYearType() { ... }
public int getYear() { ... }
public Quarter getQuarter() { ... }
public int getQuarterValue() { ... }
```

Thoughts? Suggestions on the name?

(Same question exists for `YearWeek`.)

**Comment by jodastephen** [comment]
Just noting ThreTen-Extra has `OffsetDate` which be tackled too, although that can wait until there is a known pattern for `LocaLDate`.

On the getters, there is no good answer. I hadn't appreciated you were intending to tackle those too. I don't think there is any chance of getting a `getYearObj()` method into the JDK (well at least not until inline/value types appear, which just might affect things). A such, my feeling is that adding them to `YearWeek` and `YearQuarter` is a little out of place. `getYearObj()` is my best naming suggestion.
